src/devices/GoPiGo3/GoPiGo3.cs

```diff
@@ -0,0 +1,900 @@
+                }
+            if ((GoPiGo3Info.Manufacturer != ManufacturerName) || (GoPiGo3Info.Board != BoardName))
+                throw new IOException($"GoPiGo3 with address {SpiAddress} not connected.");
+            MotorGearRatio = DefaultMotorGearRatio;
+            EncoderTicksPerRotation = DefaultEncoderTicksPerRotation;
+            // Initialise the 2 Groove sensors
+            GrooveSensor = new GrooveSensor[2];
+            GrooveSensor[0] = new GrooveSensor(GroovePort.Groove1);
+            GrooveSensor[1] = new GrooveSensor(GroovePort.Groove2);
+        }
+
+        /// <summary>
+        /// Reset everything
+        /// </summary>
+        public void Dispose()
+        {
+            // Reset groove sensors
+            SetGrooveType(GroovePort.Both, GrooveSensorType.Custom);
+            SetGrooveMode(GroovePort.Both, GrooveInputOutput.InputDigital);
+            // Turn off the motors
+            SetMotorPower(MotorPort.Both, (byte)MotorSpeed.Float);
+            // Reset the motor limits
+            SetMotorLimits(MotorPort.Both, 0, 0);
+            // Turn off the servos
+            SetServo(ServoPort.Both, 0);
+            // Turn off the LEDs
+            SetLed((byte)GoPiGo3Led.LedEyeLeft + (byte)GoPiGo3Led.LedEyeRight + (byte)GoPiGo3Led.LedBlinkerLeft + (byte)GoPiGo3Led.LedBlinkerRight, 0, 0, 0);
+        }
+
+        #endregion
+
+        #region SPI transfer
+
+        /// <summary>
+        /// Conduct a SPI transaction
+        /// </summary>
+        /// <param name="dataOut">A byte array to send.The length of the array will determine how many bytes are transferred.</param>
+        /// <returns>Returns an array of the bytes read.</returns>
+        public byte[] SpiTransferArray(byte[] dataOut)
+        {
+            byte[] result = new byte[dataOut.Length];
+            _spiDevice.TransferFullDuplex(dataOut, result);
+            return result;
+        }
+
+        /// <summary>
+        /// Read a 32 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <returns>Return the value read</returns>
+        public int SpiRead32(SpiMessageType MessageType)
+        {
+            int retVal = -1;
+            byte[] outArray = { SpiAddress, (byte)MessageType, 0, 0, 0, 0, 0, 0 };
+            byte[] reply = SpiTransferArray(outArray);
+
+            if (reply[3] == 0xA5)
+            {
+                retVal = (int)(reply[4] << 24) | (reply[5] << 16) | (reply[6] << 8) | reply[7];
+            }
+            else
+            {
+                throw new IOException($"{nameof(SpiRead32)} : no SPI response");
+            }
+
+            return retVal;
+        }
+
+        /// <summary>
+        /// Read a 16 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <returns>Return the vallue read</returns>
+        public int SpiRead16(SpiMessageType MessageType)
```

Comment:
should this return short/ushort instead?

Reply:
For consistency Reason, I prefer to leave it as an int. All the read functions returns an int. I would then have to change the SpiRead8. Now if you thing it's better to have them int16 and byte, then I'll change them.

Reply:
no strong feelings so you can leave as is